rust/lance/src/dataset/sql.rs

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use crate::datafusion::LanceTableProvider;
+use crate::Dataset;
+use arrow_array::RecordBatch;
+use datafusion::dataframe::DataFrame;
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion::prelude::SessionContext;
+use std::sync::Arc;
+
+/// A SQL builder to prepare options for running SQL queries against a Lance dataset.
+#[derive(Clone, Debug)]
+pub struct SqlBuilder {
+    /// The dataset to run the SQL query
+    pub(crate) dataset: Option<Dataset>,
+
+    /// The SQL query to run
+    pub(crate) sql: String,
+
+    /// the name of the table to register in the datafusion context
+    pub(crate) table_name: Option<String>,
+
+    /// If true, the query result will include the internal row id
+    pub(crate) row_id: bool,
+
+    /// If true, the query result will include the internal row address
+    pub(crate) row_addr: bool,
+}
+
+impl SqlBuilder {
+    /// The table name to register in the datafusion context.
+    /// This is used to specify a "table name" for the dataset.
+    /// So that you can run SQL queries against it.
+    /// If not set, the default table name is "dataset".
+    pub fn table_name(mut self, table_name: &str) -> Self {
+        self.table_name = Some(table_name.to_string());
+        self
+    }
+
+    /// Specify if the query result should include the internal row id.
+    /// If true, the query result will include an additional column named "_rowid".
+    pub fn with_row_id(mut self, row_id: bool) -> Self {
+        self.row_id = row_id;
+        self
+    }
+
+    /// Specify if the query result should include the internal row address.
+    /// If true, the query result will include an additional column named "_rowaddr".
+    pub fn with_row_addr(mut self, row_addr: bool) -> Self {
+        self.row_addr = row_addr;
+        self
+    }
+
+    pub async fn execute(self) -> lance_core::Result<QueryResult> {
```

Comment:
just one other thing, I think this should not be called `execute` but just `build`, because the SQL query has not really been fully executed at this point.